File: pepper-farm/backend/schemas/task.py
from pydantic import BaseModel, field_validator
from datetime import datetime
from typing import Optional

ALLOWED_PRIORITIES = {"low", "medium", "high", "critical"}


class AlertInfo(BaseModel):
    severity: str
    metricName: str
    actualValue: float
    minAllowed: Optional[float]
    maxAllowed: Optional[float]
    message: str
    isResolved: bool
    createdAtUtc: str

class CreateTaskRequest(BaseModel):
    title: str
    taskType: str
    description: Optional[str] = None
    priority: str = "medium"
    assignedToUserId: Optional[int] = None
    dueDate: Optional[datetime] = None
    pepperId: Optional[int] = None
    zoneId: Optional[int] = None
    zoneCode: Optional[str] = None  # alternative to zoneId; resolved server-side
    anomalyId: Optional[int] = None  # links task to the source SensorAlert


class UpdateTaskRequest(BaseModel):
    title: Optional[str] = None
    taskType: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None
    assignedToUserId: Optional[int] = None
    dueDate: Optional[datetime] = None
    zoneCode: Optional[str] = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Title is required.")
        return v

    @field_validator("taskType")
    @classmethod
    def task_type_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("TaskType is required.")
        return v

class TaskResponse(BaseModel):
    id: int
    title: str
    description: Optional[str]
    status: str
    priority: str
    taskType: str
    createdByUserId: int
    assignedToUserId: Optional[int]
    dueDate: Optional[datetime]
    startedAt: Optional[datetime]
    completedAt: Optional[datetime]
    pepperId: Optional[int]
    zoneId: Optional[int]
    zoneCode: Optional[str]
    anomalyId: Optional[int]
    alertInfo: Optional[AlertInfo] = None
    createdAt: datetime
    updatedAt: datetime

    model_config = {"from_attributes": True}

File: pepper-farm/frontend/src/types/task.ts
export type TaskPriority = 'low' | 'medium' | 'high' | 'critical';
export type TaskStatus = 'todo' | 'in_progress' | 'done' | 'cancelled';

export interface AlertInfo {
  severity: string;
  metricName: string;
  actualValue: number;
  minAllowed: number | null;
  maxAllowed: number | null;
  message: string;
  isResolved: boolean;
  createdAtUtc: string;
}

export interface ChecklistItem {
  itemId: number;
  title: string;
  isCompleted: boolean;
  position: number;
}

// Used inside TaskForm: existing items carry itemId + isCompleted; new items don't.
export interface ChecklistFormItem {
  itemId?: number;
  title: string;
  isCompleted: boolean;
}

export interface CreateTaskFormData {
  title: string;
  description: string;
  taskType: string;
  priority: TaskPriority;
  assignedToUserId: string;
  dueDate: string;
  zoneCode: string; // section id like 'GH-01', empty = none
  checklistItems: ChecklistFormItem[];
}

export interface Task {
  id: number;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  taskType: string;
  createdByUserId: number;
  assignedToUserId: number | null;
  dueDate: string | null;
  startedAt: string | null;
  completedAt: string | null;
  pepperId: number | null;
  zoneId: number | null;
  zoneCode: string | null;
  anomalyId: number | null;
  alertInfo: AlertInfo | null;
  createdAt: string;
  updatedAt: string;
  checklistItems: ChecklistItem[];
}
